### PR TITLE
go 1.11 port and TFO

### PR DESCRIPTION
**Pull request by dstiliadis**
This PR changes the TCP connection manipulations to use the go 1.11 structures (ListenerConfig and Dialer Control), and activates TFO for the proxy connections. Includes several other refactors and optimizations. Results in a 20% improvement in performance for the proxies.

It also removes the splicing proxy and delegates splicing to io.copy from Go.

### CODE REVIEW

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "780690bc7f94653473df78c32abb55b7bacdb23c"
  },
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "6320ff8566daa53ac50f9258d80afba94bed551d"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "aab2da224e85ddbb0ac6c12e85f2a4eeb2cf42bf"
  },
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "8f5cb2695a800ebce0016222705edce8f00d8f44"
  }
]
```

File: controller/internal/enforcer/applicationproxy/tcp/tcp.go [hunk 1]
```diff
@@ -268,14 +269,17 @@ func (p *Proxy) puContextFromContextID(puID string) (*pucontext.PUContext, error
 }
 
 // Initiate the downstream connection
-func (p *Proxy) downConnection(ip net.IP, port int) (net.Conn, error) {
+func (p *Proxy) downConnection(ctx context.Context, ip net.IP, port int) (net.Conn, error) {
 
+	if port == 20996 {
```
Comment (varks) [inline controller/internal/enforcer/applicationproxy/tcp/tcp.go]: leftover debug message?

File: utils/allocator/allocator.go [hunk 2]
```diff
@@ -34,5 +35,24 @@ func (p *allocator) Release(item string) {
 	if len(p.allocate) == cap(p.allocate) {
 		return
 	}
+
+	intItem, err := strconv.Atoi(item)
+	if err == nil {
```
Comment (varks) [inline utils/allocator/allocator.go]: looks like the check is inverted? Also we should check if intItem is 0 and ignore releasing it like we do for ReleaseInt.

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "8f5cb2695a800ebce0016222705edce8f00d8f44"
  },
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "aab2da224e85ddbb0ac6c12e85f2a4eeb2cf42bf"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "8f5cb2695a800ebce0016222705edce8f00d8f44"
  },
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "aab2da224e85ddbb0ac6c12e85f2a4eeb2cf42bf"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "e3280a8f37245169545316eb0b9cb43eba6f63e1"
  },
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "448087098003f917e37b36266aab51987403743f"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "apotests",
    "pr-id": "1890",
    "commit-sha": "90db76271cae0d8dab4139e18993735caf58321c"
  },
  {
    "project": "tfo",
    "component": "apotests-lib",
    "pr-id": "75",
    "commit-sha": "f09b91e78ebf0660e76e15e88e3487267b3f07cd"
  },
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "7acea472f3dda5adaaad699cd9a3fc3696b792e6"
  },
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "cab7dae9b4c7ac98b8e89965cbb6144e4075396c"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "7acea472f3dda5adaaad699cd9a3fc3696b792e6"
  },
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "2ad645100676649a70c56b0fb5667357952c5909"
  },
  {
    "project": "tfo",
    "component": "apotests",
    "pr-id": "1890",
    "commit-sha": "90db76271cae0d8dab4139e18993735caf58321c"
  },
  {
    "project": "tfo",
    "component": "apotests-lib",
    "pr-id": "75",
    "commit-sha": "f09b91e78ebf0660e76e15e88e3487267b3f07cd"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "apotests-lib",
    "pr-id": "75",
    "commit-sha": "f09b91e78ebf0660e76e15e88e3487267b3f07cd"
  },
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "fe86dd915d410a3cf81862cad2be47ea89125542"
  },
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "33fa66c1008285d9c8b82e2cc2c30be95a6ed30e"
  },
  {
    "project": "tfo",
    "component": "apotests",
    "pr-id": "1890",
    "commit-sha": "90db76271cae0d8dab4139e18993735caf58321c"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "apotests",
    "pr-id": "1890",
    "commit-sha": "90db76271cae0d8dab4139e18993735caf58321c"
  },
  {
    "project": "tfo",
    "component": "apotests-lib",
    "pr-id": "75",
    "commit-sha": "f09b91e78ebf0660e76e15e88e3487267b3f07cd"
  },
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "fe86dd915d410a3cf81862cad2be47ea89125542"
  },
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "f183d41374f24f08fbdeb2ca6ff1822c25cd6e9e"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "apotests-lib",
    "pr-id": "75",
    "commit-sha": "f09b91e78ebf0660e76e15e88e3487267b3f07cd"
  },
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "fe86dd915d410a3cf81862cad2be47ea89125542"
  },
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "6c264facac042aa805db80cae4bf4a01a87c31f7"
  },
  {
    "project": "tfo",
    "component": "apotests",
    "pr-id": "1890",
    "commit-sha": "90db76271cae0d8dab4139e18993735caf58321c"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "apotests-lib",
    "pr-id": "75",
    "commit-sha": "f09b91e78ebf0660e76e15e88e3487267b3f07cd"
  },
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "aef48fd5172b13354eeefe362a3a46c8043f38d4"
  },
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "6bdd83f6f88adb67a2a9dfd53a6d02068d5c85c9"
  },
  {
    "project": "tfo",
    "component": "apotests",
    "pr-id": "1890",
    "commit-sha": "90db76271cae0d8dab4139e18993735caf58321c"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "apotests-lib",
    "pr-id": "75",
    "commit-sha": "f09b91e78ebf0660e76e15e88e3487267b3f07cd"
  },
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "aef48fd5172b13354eeefe362a3a46c8043f38d4"
  },
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "306dcb04c603ea4c0d2fbe06637fd8fa4720696d"
  },
  {
    "project": "tfo",
    "component": "apotests",
    "pr-id": "1890",
    "commit-sha": "90db76271cae0d8dab4139e18993735caf58321c"
  }
]
```

**Comment by ericrpowers** [comment]
/build - automatically fired by gogo with following PRs and commit SHAs v1.0.0
```
[
  {
    "project": "tfo",
    "component": "enforcerd",
    "pr-id": "975",
    "commit-sha": "9b9a33901e50a133ed26bb3f5baf15d47ee19560"
  },
  {
    "project": "tfo",
    "component": "apotests",
    "pr-id": "1890",
    "commit-sha": "90db76271cae0d8dab4139e18993735caf58321c"
  },
  {
    "project": "tfo",
    "component": "apotests-lib",
    "pr-id": "75",
    "commit-sha": "f09b91e78ebf0660e76e15e88e3487267b3f07cd"
  },
  {
    "project": "tfo",
    "component": "trireme-lib",
    "pr-id": "711",
    "commit-sha": "aef48fd5172b13354eeefe362a3a46c8043f38d4"
  }
]
```

**Pull request merged**